packages/NodeTypeResolver/PHPStan/Scope/NodeVisitor/StmtKeyNodeVisitor.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\NodeTypeResolver\PHPStan\Scope\NodeVisitor;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt;
+use PhpParser\Node\Stmt\ClassLike;
+use PhpParser\NodeVisitorAbstract;
+use Rector\Core\Contract\PhpParser\Node\StmtsAwareInterface;
+use Rector\NodeTypeResolver\Node\AttributeKey;
+use Rector\NodeTypeResolver\PHPStan\Scope\Contract\NodeVisitor\ScopeResolverNodeVisitorInterface;
+
+final class StmtKeyNodeVisitor extends NodeVisitorAbstract implements ScopeResolverNodeVisitorInterface
+{
+    public function enterNode(Node $node): ?Node
+    {
+        // need direct Stmt instance check to got every Stmt
+        if (! $node instanceof Stmt || $node instanceof ClassLike) {
+            return null;
+        }
+
+        // re-index stmt key under current node
+        if ($node->getAttribute(AttributeKey::STMT_KEY) !== null) {
+            $this->setStmtKeyAttribute($node);
+            return null;
+        }
+
+        $parentNode = $node->getAttribute(AttributeKey::PARENT_NODE);
+        if (! $parentNode instanceof StmtsAwareInterface) {
+            return null;
+        }
```

Comment:
I think this visitor needs to work without accessing the parent node.

couldn't you just save the parent in a local property instead?

Reply:
'parent' attribute is essential to reindex the key under it. Stmt may just added or removed, I need to cover `->addNodeBeforeNode()` and `->addNodeAfterNode()` or return array of Stmt in `refactor()`

Reply:
I don't say you should n't work with the parent element, but I think you could get a reference to the parent element without using `$node->getAttribute(AttributeKey::PARENT_NODE);`.

---

or do you mean its not realistic that we can get rid of `$node->getAttribute(AttributeKey::PARENT_NODE);` in the long run and therefore it doesn't matter if we have this call-site here or not?

Reply:
I am thinking that `PARENT_NODE` can be reduced, but not to be removed since it used to resolve the Scope. We need to have a "SlimNodeConnectingVisitor" which only fill "parent" attribute without prev or next in the end.